### PR TITLE
Dynamic Recipes

### PR DESCRIPTION
**Pull request by ZephireNZ**
This provides the implementation side to GlowstoneMC/Glowkit#34. It makes banners, with others yet to come, craftable without creating countless entries in the static file.

Recipes to implement:
- [x] Banners
  - [x] Layer adding
  - [x] Duplication
- [x] Fireworks - Requires #506 for implementation
  - [x] Stars creation/editing
  - [x] Firework creation
- [x] Book duplication
- [x] Tool/Armor repair
- [x] Armor dyeing - Requires #496 for implementation
- [ ] Maps
  - [ ] Map zooming
  - [x] Map duplication

### CODE REVIEW

**Comment by turt2live** [comment]
@ZephireNZ Something really strange happened to your commit history here. You may want to consider looking into it.

**Comment by ZephireNZ** [comment]
@turt2live fixed. Somehow my rebase fucked up royally. Should be all good now. I even purged all the banner stuff (although it will need to be built with it)

**Comment by turt2live** [comment]
Interesting, my bot really screwed up here.

**Comment by turt2live** [comment]
You know what @turt2bot. Screw you.

**Comment by ZephireNZ** [comment]
@turt2live: Your bot ain't perfect :P Of course there's going to be changes in a WIP PR.

**Comment by turt2live** [comment]
I've edited the description because @dequis's script fails with multiple links. Glowstone/Glowkit#24 -> Glowkit#24.

**Comment by ZephireNZ** [comment]
I've removed that line completely, now that Banners has been merged.

**Comment by turt2live** [comment]
Fair point, thanks!

**Comment by dequis** [comment]
Welp, looks like you forgot ItemMatcher in the glowkit PR.

Better wait until @ZephireNZ's computer unfucks itself, I guess?

**Comment by ZephireNZ** [comment]
Uhh... Balls, looks like it. Not sure how I missed it, but oh well. Will be
out of action for about a week and a half.

**Comment by dequis** [comment]
Meh, no big deal, turt2live has a big enough bunch of PRs to test meanwhile.

**Comment by ZephireNZ** [comment]
@turt2live Would you be able to have a look over the map zoom recipe, and tell me whether I should try implement the zooming using the Bukkit API? Or perhaps I should just leave that for when maps are implemented.

I'm not all sure that it's possible given that maps are almost completely unimplemented at this point.

File: src/main/java/net/glowstone/block/ItemTable.java [hunk 1]
```diff
@@ -194,6 +190,51 @@ private void registerBuiltins() {
...
+        reg(Material.FLINT_AND_STEEL, new ItemDamageable(65));
+        reg(Material.FISHING_ROD, new ItemDamageable(65));
+        reg(Material.SHEARS, new ItemDamageable(239));
+        reg(Material.BOW, new ItemDamageable(385));
+        reg(Material.WOOD_HOE, new ItemHoe(59));
+        reg(Material.STONE_HOE, new ItemHoe(131));
+        reg(Material.IRON_HOE, new ItemHoe(250));
+        reg(Material.GOLD_HOE, new ItemHoe(33));
+        reg(Material.DIAMOND_HOE, new ItemHoe(1561));
+        for (Material mat : Arrays.asList(Material.GOLD_PICKAXE, Material.GOLD_AXE, Material.GOLD_SPADE, Material.GOLD_SWORD)) {
```
Comment (bendem) [inline src/main/java/net/glowstone/block/ItemTable.java]: Why are you using a list? Would it make sense to simply use an array? (That's maybe a dumb question)

Comment (ZephireNZ) [inline src/main/java/net/glowstone/block/ItemTable.java]: Would that really be any more productive? All Arrays.asList does is wrap the array created with the args. This is arguably more readable as well.

**Comment by Tonodus** [comment]
@ZephireNZ Why don't you use Material#getMaxDurability()?

**Comment by ZephireNZ** [comment]
aaaand I'm blind -_- Thank you for that.

**Comment by dequis** [comment]
Maybe consider removing this from [WIP] so turt2live can start looking at it?

**Comment by ZephireNZ** [comment]
First have to use @Tonodus' suggestion and get rid of all the stuff in ItemTable.

**Comment by gdude2002** [comment]
This doesn't build here despite Travis being totally happy to build it: http://bamboo.gserv.me/browse/GLOW-SRV-PR445-125

I'm assuming this requires a Glowkit, though.

[log](http://i.turt2live.com/kQjW.txt)

---

Edited by turt2live: Paste sites are awesome.

**Comment by dequis** [comment]
It does, the one linked in the description. You can call `./etc/fetch_glowkit_pr.sh 445` from your scripts to fetch and install the same glowkit as travis automatically (can ignore rate limit warnings unless you build 30+ PRs per hour)

**Comment by gdude2002** [comment]
@dequis Is there a way to easily remove that installed glowkit after?

EDIT: Bearing in mind I might not always know where it's been installed to

**Comment by dequis** [comment]
Bro do you even sandbox

But yeah sure whatever, remove `~/.m2/repository/net/glowstone/glowkit/`.

[Stack overflow told me it's ok](http://stackoverflow.com/a/15358977/2195033)

**Comment by dequis** [comment]
Referencing #438 (fixes it)

**Comment by ZephireNZ** [comment]
@turt2live Alright, this is ready for review. I haven't implemented map creation simply because that is not yet implemented in Glowstone (and would be too painful to implement myself).

This currently does not selectively clear the matrix, for example when cloning an item. However, this is a problem for recipes across the board, so is out of the scope of this PR.

File: src/main/java/net/glowstone/inventory/GlowCraftingInventory.java [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package net.glowstone.inventory;
 
 import net.glowstone.GlowServer;
+import net.glowstone.inventory.crafting.CraftingManager;
```
Comment (turt2live) [inline src/main/java/net/glowstone/inventory/GlowCraftingInventory.java:4]: ... but why?

File: src/main/java/net/glowstone/inventory/crafting/GlowArmorDyeMatcher.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.LeatherArmorMeta;
+import org.bukkit.material.Dye;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class GlowArmorDyeMatcher extends ItemMatcher {
+
+    private static final List<Material> LEATHERS = ImmutableList.of(Material.LEATHER_HELMET,
```
Comment (turt2live) [inline src/main/java/net/glowstone/inventory/crafting/GlowArmorDyeMatcher.java]: We have a material matcher, please consider using it

**Comment by turt2live** [comment]
@ZephireNZ Your pull request has been unmergeable for 3 days now. Please resolve the conflicts as soon as possible, thanks.

File: src/main/java/net/glowstone/inventory/GlowCraftingInventory.java (hunk 2 above)
Comment (ZephireNZ) [inline src/main/java/net/glowstone/inventory/GlowCraftingInventory.java:4]: Moved crafting stuff into its own package. As a result, it's not in the same package as the inventory (and has to be imported)

**Pull request closed without merging**